dev/bots/utils.dart

```diff
@@ -569,9 +570,26 @@ List<T> selectIndexOfTotalSubshard<T>(List<T> tests, {String subshardKey = kSubs
     return <T>[];
   }
 
-  final int testsPerShard = (tests.length / total).ceil();
-  final int start = (index - 1) * testsPerShard;
-  final int end = math.min(index * testsPerShard, tests.length);
+  // While there exists a closed formula figuring out the range of tests this
+  // shard is resposible for, modeling this as a simulation of distributing
+  // items equally into buckets is more intuitive.
+  //
+  // A bucket represents how many tests a shard should be allocated.
+  final List<int> buckets = List<int>.filled(total, 0);
+  // First, allocate an even amount of items to each bucket.
+  for (int i = 0; i < buckets.length; i++) {
+    buckets[i] = (tests.length / total).floor();
+  }
+  // For the N leftover items, put one into each of the first N buckets.
+  final int remainingItems = tests.length % buckets.length;
+  for (int i = 0; i < remainingItems; i++) {
+    buckets[i] += 1;
+  }
+
+  // Lastly, compute the indices of the items in buckets[index].
+  final int numberOfItemsInPreviousBuckets = index == 0 ? 0 : buckets.sublist(0, index - 1).sum;
+  final int start = (numberOfItemsInPreviousBuckets + 1) - 1;
```

Comment:
Can we simplify this to `final int start = numberOfItemsInPreviousBuckets;`?

Which is more intuitive is maybe subjective, but to me this is just more confusing.

Reply:
Did so in [refactor: remove one-indexing adjustment](https://github.com/flutter/flutter/pull/156768/commits/247303d5379b28c2c5c30d03f7cd06b2eb9a294c)